src/wasm/wasm-type.cpp

```diff
@@ -1347,13 +1635,304 @@ Type TypeBuilder::getTempRttType(size_t i, uint32_t depth) {
+        // If the split partition was a potential distinguisher, both smaller
+        // partitions are as well. Otherwise, we only need to add the smaller of
+        // the two smaller partitions as a new potential distinguisher.
+        if (distinguishers.count(distinguishedIndex) ||
+            partitions[newPartitionIndex].size() <=
+              partitions[distinguishedIndex].size()) {
+          distinguishers.insert(newPartitionIndex);
+        } else {
+          distinguishers.insert(distinguishedIndex);
+        }
+      }
+    }
+  }
+
+  translatePartitionsToTypes();
+}
+
+void ShapeCanonicalizer::initializePredecessors() {
+  for (auto heapType : input) {
+    size_t childIndex = 0;
+    for (auto* child : getChildren(heapType)) {
+      alphabetSize = std::max(alphabetSize, childIndex + 1);
+      preds[*child][childIndex++].insert(heapType);
+    }
+  }
+}
+
+void ShapeCanonicalizer::initializePartitions() {
+  // Create the initial partitions based on the top-level shape of the input
+  // heap types. If two heap types are differentiable without recursing into
+  // their child heap types, then they are obviously not equivalent and can be
+  // placed in different partitions. Starting with this fine-grained partition
+  // lets us use simple child indices as our transition alphabet since we will
+  // never mix up equivalent indices from different kinds of types, for example
+  // considering a struct and a signature with the same children to be the same
+  // type.
+  std::unordered_map<ShallowHeapType, size_t> initialIndices;
+  for (auto type : input) {
+    ShallowHeapType shallow(type);
+    auto inserted = initialIndices.insert({shallow, partitions.size()});
+    if (inserted.second) {
+      // We have not seen a type with this shape before; create a new
+      // partition.
+      partitionIndices[type] = partitions.size();
+      partitions.emplace_back(TypeSet{type});
+    } else {
+      // Add to the partition we have already created for this type shape.
+      size_t index = inserted.first->second;
+      partitionIndices[type] = index;
+      partitions[index].insert(type);
+    }
+  }
+}
+
+void ShapeCanonicalizer::translatePartitionsToTypes() {
+  // Create a single new HeapTypeInfo for each partition. Initialize each new
+  // HeapTypeInfo as a copy of a representative HeapTypeInfo from its partition,
+  // then patch all the children of the new HeapTypeInfos to refer to other new
+  // HeapTypeInfos rather than the original HeapTypeInfos. This newly formed
+  // graph will have a shape coinductively equivalent to the original graph's
+  // shape, but each type definition will be minimal and distinct.
+  for (auto& partition : partitions) {
+    const auto& representative = *getHeapTypeInfo(*partition.begin());
+    infos.push_back(std::make_unique<HeapTypeInfo>(representative));
+    infos.back()->isTemp = true;
+  }
+  for (auto& info : infos) {
+    for (auto* child : getChildren(asHeapType(info))) {
```

Comment:
I don't understand this loop. Each input HeapType is in one of the `partitions`, represented by one of the `infos` (at the same index). Can't we traverse all the HeapTypes by traversing the partitions and then the partition's contents? Traversing the children I suppose will cover the same things in the end, but it seems like more work?

(then `getChildren` would only be used in the predecessor calculation, which is what I'd expect is the main place that actually looks at the shape details)

Reply:
We don't just need the HeapTypes; we need the HeapType _locations_ where they are reachable from the other HeapTypes so we can replace their uses in-place with uses of canonical HeapTypes.

Reply:
I see, thanks. Makes sense. I guess that's already explained in the comment above but I missed it.